### PR TITLE
[WIP] Release version 0.2.0

### PR DESCRIPTION
**Pull request by dokato**
A lot of changes from @agwells and me. Details in commits. The most important include:
- changing core of router from `page.js` to `updateQueryString`
- debugging function `log_msg` added
- now it's possible to catch get params
- change in a way of link declaration (by `route_link`)

### CODE REVIEW

File: R/misc.R [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  slicefrom <- 1
+  if (substr(hashpath, slicefrom, slicefrom) == "#") {
+    slicefrom <- slicefrom + 1
+  }
+  if (substr(hashpath, slicefrom, slicefrom) == "!") {
+    slicefrom <- slicefrom + 1
+  }
+  if (substr(hashpath, slicefrom, slicefrom) == "/") {
+    slicefrom <- slicefrom + 1
+  }
```
Comment (przytu1) [inline R/misc.R]: I imagine there are a lot of corner cases resistant for these 3 if statements. E.g. `##!//url` is still valid part of the url for the browser. How about something like this:

```
slicefrom <- 1
while (substr(hashpath, slicefrom, slicefrom) %in% c("#", "!", "/")) slicefrom <- slicefrom + 1
```

What do you think? Also you have one typo in the comment above ("hashbang path morker" <-- "o")

File: R/misc.R [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+#' Extract link name
+#'
+#' Strips off the first 3 character, assuming that they are: "#!/".
+#'
+#' @param path character with link path
+#' @return stripped link
+#' @example
+#' extract_link_name("#!/abc")
+extract_link_name <- function(path) {
```
Comment (przytu1) [inline R/misc.R]: Do you assume `path` is always correct? How about:
```
extract_link_name <- function(path) {
 sub("#!/", "", cleanup_hashpath(path))
}
```

(`sub` replaces only the first occurence of a pattern)

File: R/misc.R [hunk 3]
```diff
@@ -0,0 +1,73 @@
+#' Helper function to print out log messages into Shiny
```
Comment (przytu1) [inline R/misc.R]: I am not a big fan of modules named "misc", "utils", etc. This doesn't give me any additional information when I look for some function. I would suggest to rename it to `logger.R` and keep `log_msg` only, and move `cleanup_hashpath` and `extract_link_name` to the `router.R`

File: examples/basic/app.R [hunk 4]
```diff
@@ -5,7 +5,7 @@ library(shiny.router)
 menu <- (
   tags$ul(
     tags$li(a(class = "item", href = "/", "Page")),
-    tags$li(a(class = "item", href = "/other", "Other page"))
+    tags$li(a(class = "item", href = route_link("other"), "Other page"))
```
Comment (krystian8207) [inline examples/basic/app.R]: Please export `route_link` function or use `shiny.router:::route_link` instead.

File: R/router.R [hunk 5]
```diff
@@ -107,16 +130,105 @@ make_router <- function(default, ...) {
...
+#' @param session The Shiny session
+#' @return The full list of params on the URL (if any), as a list. Or, the single
+#' requested param (if present). Or FALSE if there's no input, or no params.
+#' @reactivesource
+#' @export
+get_query_param <- function(field = NULL, session = shiny::getDefaultReactiveDomain()) {
+  log_msg("Trying to fetch field '", field)
+
+  if (is.null(session$userData$shiny.router.page()$query)) {
+    return(FALSE)
```
Comment (krystian8207) [inline R/router.R]: I suppose `FALSE` output is not intuitive here. I other cases there is returned string, so maybe we could return NULL in here?

File: R/misc.R (hunk 1 above)
Comment (dokato) [reply R/misc.R]: Thanks, that's useful!

File: R/misc.R (hunk 2 above)
Comment (dokato) [reply R/misc.R]: Good idea, changed.

File: examples/basic/app.R (hunk 4 above)
Comment (dokato) [reply examples/basic/app.R]: I fixed semantic example now.

File: R/router.R (hunk 5 above)
Comment (dokato) [reply R/router.R]: 👍

File: examples/basic/app.R (hunk 4 above)
Comment (dokato) [reply examples/basic/app.R]: and NAMESPACE updated

**Pull request merged**